src/iceberg/catalog.h

```diff
@@ -0,0 +1,178 @@
+  /// \return a Table instance or Error::kAlreadyExists if the table already exists
+  virtual expected<std::shared_ptr<Table>, ErrorKind> RegisterTable(
+      const TableIdentifier& identifier, const std::string& metadata_file_location) = 0;
+
+  /// \brief Initialize a catalog given a custom name and a map of catalog properties
+  ///
+  /// A custom Catalog implementation must have a default constructor. A compute engine
+  /// will first initialize the catalog without any arguments, and then call this method
+  /// to complete catalog initialization with properties passed into the engine.
+  ///
+  /// \param name a custom name for the catalog
+  /// \param properties catalog properties
+  virtual void Initialize(const std::string& name,
+                          const std::map<std::string, std::string>& properties) = 0;
+
+  /// \brief Instantiate a builder to either create a table or start a create/replace
+  /// transaction
+  ///
+  /// \param identifier a table identifier
+  /// \param schema a schema
+  /// \return the builder to create a table or start a create/replace transaction
+  virtual std::unique_ptr<class TableBuilder> BuildTable(
+      const TableIdentifier& identifier, const Schema& schema) const = 0;
+
+  /// \brief A builder used to create valid tables or start create/replace transactions
+  class TableBuilder {
+   public:
+    virtual ~TableBuilder() = default;
+
+    /// \brief Sets a partition spec for the table
+    ///
+    /// \param spec a partition spec
+    /// \return this for method chaining
+    virtual TableBuilder& WithPartitionSpec(const PartitionSpec& spec) = 0;
+
+    /// \brief Sets a sort order for the table
+    ///
+    /// \param sort_order a sort order
+    /// \return this for method chaining
+    virtual TableBuilder& WithSortOrder(const SortOrder& sort_order) = 0;
+
+    /// \brief Sets a location for the table
+    ///
+    /// \param location a location
+    /// \return this for method chaining
+    virtual TableBuilder& WithLocation(const std::string& location) = 0;
+
+    /// \brief Adds key/value properties to the table
+    ///
+    /// \param properties key/value properties
+    /// \return this for method chaining
+    virtual TableBuilder& WithProperties(
+        const std::map<std::string, std::string>& properties) = 0;
+
+    /// \brief Adds a key/value property to the table
+    ///
+    /// \param key a key
+    /// \param value a value
+    /// \return this for method chaining
+    virtual TableBuilder& WithProperty(const std::string& key,
+                                       const std::string& value) = 0;
+
+    /// \brief Creates the table
+    ///
+    /// \return the created table
+    virtual std::unique_ptr<Table> Create() = 0;
+
+    /// \brief Starts a transaction to create the table
+    ///
+    /// \return the Transaction to create the table
+    virtual std::unique_ptr<Transaction> CreateTransaction() = 0;
```

Comment:
Just a thought. To make this more Iceberg REST catalog agnostic, we could also call this `Stage`:

https://github.com/apache/iceberg/blob/a4816c1c99063770473920cd6d62f88f90a292dc/open-api/rest-catalog-open-api.yaml#L556-L561

Reply:
Quick question: `Transaction` has a comment `A transaction for performing multiple updates to a table` but the `BaseTransaction` has a check `checkLastOperationCommitted` which throws when previous update is uncommitted. Will multiple updates be supported in the future to remove this inconsistency?

Reply:
> Will multiple updates be supported in the future to remove this inconsistency?

Yes, there is no reason not to support this.